File: backend/api/project.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List
from backend.database import get_db
from backend.model.models import Project
from backend.api.project_schemas import ProjectCreate, ProjectUpdate, ProjectResponse

router = APIRouter(prefix="/api/projects", tags=["Projects"])


@router.post("", response_model=ProjectResponse)
def create_project(data: ProjectCreate, db: Session = Depends(get_db)):
    existing = db.query(Project).filter(Project.name == data.name).first()
    if existing:
        raise HTTPException(status_code=400, detail="Project name already exists")
    project = Project(**data.model_dump())
    db.add(project)
    db.commit()
    db.refresh(project)
    return project


@router.get("", response_model=List[ProjectResponse])
def list_projects(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    return db.query(Project).offset(skip).limit(limit).all()


@router.get("/{project_id}", response_model=ProjectResponse)
def get_project(project_id: int, db: Session = Depends(get_db)):
    project = db.query(Project).filter(Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return project


@router.put("/{project_id}", response_model=ProjectResponse)
def update_project(project_id: int, data: ProjectUpdate, db: Session = Depends(get_db)):
    project = db.query(Project).filter(Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    for key, value in data.model_dump(exclude_unset=True).items():
        setattr(project, key, value)
    db.commit()
    db.refresh(project)
    return project


@router.delete("/{project_id}")
def delete_project(project_id: int, db: Session = Depends(get_db)):
    project = db.query(Project).filter(Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    db.delete(project)
    db.commit()
    return {"message": "Project deleted successfully"}

File: frontend/src/api/project.ts
import request from '@/utils/request'

export interface Project {
  id: number
  name: string
  description?: string
  base_url?: string
  webhook_secret?: string
  notification_config?: Record<string, any>
  created_at: string
  updated_at?: string
}

export const getProjects = (skip = 0, limit = 100) => {
  return request.get<any, Project[]>('/projects', { params: { skip, limit } })
}

export const getProject = (id: number) => {
  return request.get<any, Project>(`/projects/${id}`)
}

export const createProject = (data: Partial<Project>) => {
  return request.post<any, Project>('/projects', data)
}

export const updateProject = (id: number, data: Partial<Project>) => {
  return request.put<any, Project>(`/projects/${id}`, data)
}

export const deleteProject = (id: number) => {
  return request.delete(`/projects/${id}`)
}
